x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -60,40 +73,143 @@ public void triggered(SchedulerEngine.Event event) {
+                        deleteSnapshots(snapshotsToBeDeleted);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        running.set(false);
+                    }
+                }, err -> running.set(false));
 
             } finally {
                 running.set(false);
             }
         } else {
-            logger.debug("snapshot lifecycle retention task started, but a task is already running, skipping");
+            logger.trace("snapshot lifecycle retention task started, but a task is already running, skipping");
         }
     }
 
     static Map<String, SnapshotLifecyclePolicy> getAllPoliciesWithRetentionEnabled(final ClusterState state) {
-        // TODO: fill me in
-        return Collections.emptyMap();
+        final SnapshotLifecycleMetadata snapMeta = state.metaData().custom(SnapshotLifecycleMetadata.TYPE);
+        if (snapMeta == null) {
+            return Collections.emptyMap();
+        }
+        return snapMeta.getSnapshotConfigurations().entrySet().stream()
+            .filter(e -> e.getValue().getPolicy().getRetentionPolicy() != null)
+            .filter(e -> e.getValue().getPolicy().getRetentionPolicy().equals(SnapshotRetentionConfiguration.EMPTY) == false)
+            .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().getPolicy()));
     }
 
-    static boolean snapshotEligibleForDeletion(SnapshotInfo snapshot, Map<String, SnapshotLifecyclePolicy> policies) {
-        // TODO: fill me in
-        return false;
+    static boolean snapshotEligibleForDeletion(SnapshotInfo snapshot, List<Tuple<String, SnapshotInfo>> allSnapshots,
+                                               Map<String, SnapshotLifecyclePolicy> policies) {
+        if (snapshot.userMetadata() == null) {
+            // This snapshot has no metadata, it is not eligible for deletion
+            return false;
+        }
+
+        final String policyId;
+        try {
+            policyId = (String) snapshot.userMetadata().get(SnapshotLifecyclePolicy.POLICY_ID_METADATA_FIELD);
+        } catch (Exception e) {
+            logger.error("unable to retrieve policy id from snapshot metadata [" + snapshot.userMetadata() + "]", e);
+            throw e;
+        }
+
+        SnapshotLifecyclePolicy policy = policies.get(policyId);
+        if (policy == null) {
+            // This snapshot was taking by a policy that doesn't exist, so it's not eligible
+            return false;
+        }
+
+        SnapshotRetentionConfiguration retention = policy.getRetentionPolicy();
+        if (retention == null || retention.equals(SnapshotRetentionConfiguration.EMPTY)) {
+            // Retention is not configured
+            return false;
+        }
+
+        final String repository = policy.getRepository();
+        // Retrieve the predicate based on the retention policy, passing in snapshots pertaining only to *this* policy and repository
```

Comment:
This is likely to confuse some users (because you can change the repository for a policy, and then the snapshots in the old repo, which still have the same `policy` set, will never get aged out), but probably just needs clear docs?

Reply:
I agree, not really an easy way to change this, I think documentation is the best fix (docs for retention though we can work on in a subsequent PR)